### PR TITLE
Methods to fetch vaults from multicall

### PR DESCRIPTION
**Pull request by solovev**

### CODE REVIEW

File: app/api/endpoints.ts [hunk 1]
```diff
@@ -25,3 +26,16 @@ export const GraphQLEndpoints: GraphQLEndpointsMap = {
...
+  if (!isOnServer())
+    return '' // will use relative path on client
+
+  if (!VERCEL_ENV)
+    return 'http://localhost:3000' // local
+
+  if (VERCEL_ENV === 'production')
+    return 'https://eonian.finance' // production
+
+  return process.env.VERCEL_URL ?? '' // preview
```
Comment (LeoVS09) [inline app/api/endpoints.ts:40]: Can we use for API requests only relative URLs? Because this vercel host is much more complicated in reality.

I have high suspicion that it should be more like that
```suggestion
  if (VERCEL_ENV === 'production')
    return VERCEL_PROJECT_PRODUCTION_URL || NEXT_PUBLIC_APP_URL 
 || 'https://eonian.finance' // production

  return VERCEL_BRANCH_URL || VERCEL_URL ?? '' // preview
```

We had few issues with mirrors for metadata rendering for different vercel env that hard to spot directly, so I made investigation and full host building should be more like this https://github.com/eonian-core/web/blob/main/app/layout.tsx#L66. 
In some cases VERCEL_URL is not correct for some previews and `eonian.finance` will work in general for production, but we have `eonian.io` that I trying to keep able to override in emergency case through `NEXT_PUBLIC_APP_URL`, but it also good to keep working in parallel without override through `VERCEL_PROJECT_PRODUCTION_URL`.

So... please check is it possible to keep all request URLs relative, if not than please check this variables description in doc. Maybe I wrote accurate solution for metadata, but it not fully accurate for API

File: app/api/protocol/vaults/multicall/fetch-assets-via-multicall.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  const provider = new JsonRpcProvider(getRPCEndpoint(chainId))
+  const multicall = new Multicall(multicallAddress, provider, requests)
+  const responses = await multicall.makeRequest()
+
+  // Convert raw multicall responses to the intermediate asset model
+  const intermediateAssetModels = _.chain(responses)
+    .chunk(createAssetRequests('-').length)
+    .map((responses, index) => mapResponseToIntermediateAssetModel(assetAddresses[index], responses))
+    .compact()
+    .value()
```
Comment (LeoVS09) [inline app/api/protocol/vaults/multicall/fetch-assets-via-multicall.ts:44]: Why everywhere lodash? Looks like code from 90x, it really hard to follow. Why not write it in regular way? You anyway hardcoded everywhere that it three specific requests per asset
```suggestion

 const data = response.map(({isSuccess, data}) => isSuccess ? data : null)
 // this part I would move to another function 
 const  intermediateAssetModels = []
 for(let i = 0; i += 3; i < requests.length) {
  const  name = data[i]
  const  symbol = data[i + 1]
  const  decimals = data[i + 2]
  const address = assetAddresses[i / 3]
  
  if(![name, symbol, decimals].includes(v => v === null))
   intermediateAssetModels.push({
      name: String(name),
      symbol: String(symbol),
      decimals: Number(decimals),
      address,
  })
 }
```

File: app/api/protocol/vaults/multicall/fetch-assets-via-multicall.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+function createAssetToken(intermediateAssetModel: IntermediateAssetModel, prices: Partial<Record<TokenSymbol, Price>>): Token {
+  return {
+    id: intermediateAssetModel.symbol,
+    address: intermediateAssetModel.address,
+    decimals: intermediateAssetModel.decimals,
+    name: intermediateAssetModel.name,
+    symbol: intermediateAssetModel.symbol,
+    price: prices[intermediateAssetModel.symbol as TokenSymbol]!,
+  }
+}
```
Comment (LeoVS09) [inline app/api/protocol/vaults/multicall/fetch-assets-via-multicall.ts:69]: All good, but just for future, it much easier to read
```suggestion
function createAssetToken({
symbol,
address,
decimals,
name,
}: IntermediateAssetModel, 
prices: Partial<Record<TokenSymbol, Price>>): Token {
  return {
    id: symbol,
    address,
    decimals,
    name,
    symbol,
    price: prices[symbol  as TokenSymbol]!,
  }
}
```

File: app/api/protocol/vaults/multicall/fetch-assets-via-multicall.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+ * Fetches price data from CoinGecko.
+ */
+async function fetchPrices(intermediateAssetModels: IntermediateAssetModel[]): Promise<Partial<Record<TokenSymbol, Price>>> {
+  const symbols = intermediateAssetModels.map(asset => asset.symbol) as TokenSymbol[]
+  symbols.forEach((symbol) => {
+    if (!TokenOrder.includes(symbol))
+      throw new Error(`Cannot fetch price for unknown symbol: ${symbol}`)
+  })
+  const price = await getCurrentPriceOfTokens(symbols)
+  return _.mapValues(price, (value, tokenSymbol): Price => {
```
Comment (LeoVS09) [inline app/api/protocol/vaults/multicall/fetch-assets-via-multicall.ts:81]: You probably mean
```suggestion
  return Object.values(price).map((value, tokenSymbol): Price => {
```

File: app/api/protocol/vaults/multicall/fetch-vaults-via-multicall.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  const provider = new JsonRpcProvider(getRPCEndpoint(chainId))
+  const multicall = new Multicall(multicallAddress, provider, requests)
+  const responses = await multicall.makeRequest()
+
+  // Map sequence of the raw responses to intermediate vaults state
+  const intermediateVaultModels = _.chain(responses)
+    .chunk(createVaultRequests('-').length)
+    .map((responses, index) => mapResponseToIntermediateVaultModel(vaults[index], responses))
+    .compact()
+    .value()
```
Comment (LeoVS09) [inline app/api/protocol/vaults/multicall/fetch-vaults-via-multicall.ts:55]: Again this lodash

File: app/shared/web3/interest-rate-to-apy.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+export function convertInterestRateToAPY(interestRatePerBlock: bigint, decimals: number, blocksPerDay: number): bigint {
```
Comment (LeoVS09) [inline app/shared/web3/interest-rate-to-apy.ts:1]: We have this file for such code
https://github.com/eonian-core/web/blob/main/app/finances/apy.tsx

**Review by LeoVS09 — CHANGES_REQUESTED** [review]
Let's avoid using lodash in such amounts, espesially for business logic flow. 
I can live with it if it imported in small amounts where it exactly need per path, like this `import { reduce } from 'lodash/reduce'`. But now majority stuff that can be written in one line by regular JS is by some reason replaced with lodash

File: app/api/protocol/vaults/multicall/fetch-assets-via-multicall.ts (hunk 4 above)
Comment (solovev) [reply app/api/protocol/vaults/multicall/fetch-assets-via-multicall.ts:81]: what? absolutely not, i don't need an array

File: app/api/protocol/vaults/multicall/fetch-vaults-via-multicall.ts (hunk 5 above)
Comment (solovev) [reply app/api/protocol/vaults/multicall/fetch-vaults-via-multicall.ts:55]: cool!

File: app/api/protocol/vaults/multicall/fetch-assets-via-multicall.ts (hunk 3 above)
Comment (solovev) [reply app/api/protocol/vaults/multicall/fetch-assets-via-multicall.ts:69]: matter of taste i would say

File: app/api/coin-gecko/price/get-current-price.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  return `https://api.coingecko.com/api/v3/simple/price?ids=${encodeURIComponent(ids)}&vs_currencies=usd`
+}
+
+export async function getPriceOfTokens(symbols: TokenSymbol[], revalidate: number): Promise<TokenToPrice> {
+  const url = createURL(symbols)
+  const result = await makeCoinGeckoRequest<Record<string, { usd: number }>>(url, revalidate)
+  return _.chain(result)
+    .mapKeys((_, key) => coinGeckoIdToToken[key])
+    .mapValues(value => value.usd)
+    .value()
```
[the remark this answers is not in the document]
Comment (solovev) [reply app/api/coin-gecko/price/get-current-price.ts:19]: discussed in private, decided to leave it for now

File: app/api/coin-gecko/token-symbol-to-id.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+export const tokenToCoinGeckoId: Record<TokenSymbol, string> = {
+  BTCB: 'bitcoin',
+  ETH: 'ethereum',
+  USDT: 'tether',
+  USDC: 'usd-coin',
+  DAI: 'dai',
+  BNB: 'binancecoin',
+}
+
+export const coinGeckoIdToToken = _.invert(tokenToCoinGeckoId) as Record<string, TokenSymbol>
```
[the remark this answers is not in the document]
Comment (solovev) [reply app/api/coin-gecko/token-symbol-to-id.ts:13]: discussed in private, decided to leave it for now

File: app/api/protocol/vaults/multicall/fetch-assets-via-multicall.ts (hunk 2 above)
Comment (solovev) [reply app/api/protocol/vaults/multicall/fetch-assets-via-multicall.ts:44]: discussed in private, decided to leave it for now

File: app/api/endpoints.ts (hunk 1 above)
Comment (solovev) [reply app/api/endpoints.ts]: done

**Review by LeoVS09 — APPROVED** [review]
